File: app/src/main/java/com/bridou_n/beaconscanner/features/beaconList/BeaconsRecyclerViewAdapter.java
package com.bridou_n.beaconscanner.features.beaconList;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bridou_n.beaconscanner.R;
import com.bridou_n.beaconscanner.models.BeaconSaved;
import com.bridou_n.beaconscanner.utils.CountHelper;

import java.text.DateFormat;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.realm.OrderedRealmCollection;
import io.realm.RealmRecyclerViewAdapter;

/**
 * Created by bridou_n on 30/09/2016.
 */

public class BeaconsRecyclerViewAdapter extends RealmRecyclerViewAdapter<BeaconSaved, BeaconsRecyclerViewAdapter.BaseHolder> {
    private static final String     TAG = "BEACONS_RV_ADAPTER";
    private static final int        VIEW_TYPE_EDDYSTONE_UID = 1;
    private static final int        VIEW_TYPE_EDDYSTONE_URL = 2;
    private static final int        VIEW_TYPE_IBEAON_ALTBEACON = 3;
    private static final int        VIEW_TYPE_FOOTER = 4;
    private int                     expandedPosition = -1;

    private Context ctx;

    public BeaconsRecyclerViewAdapter(@NonNull Context context, @Nullable OrderedRealmCollection<BeaconSaved> data, boolean autoUpdate) {
        super(context, data, autoUpdate);
        ctx = context;
    }

    public static class BaseHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.more_info) View moreInfo;
        @BindView(R.id.beacon_type) TextView beaconType;
        @BindView(R.id.address) TextView address;
        @BindView(R.id.distance) TextView distance;
        @BindView(R.id.distance_qualifier) TextView distanceQualifier;
        @BindView(R.id.rssi) TextView rssi;
        @BindView(R.id.tx) TextView tx;
        @BindView(R.id.manufacturer) TextView manufacturer;
        @BindView(R.id.tlm_data) LinearLayout tlmData;
        @BindView(R.id.battery) TextView battery;
        @BindView(R.id.ticks) TextView pduCount;
        @BindView(R.id.uptime) TextView uptime;
        @BindView(R.id.temperature) TextView temperature;
        @BindView(R.id.last_seen) TextView lastSeen;
        BeaconsRecyclerViewAdapter adapter;

        public BaseHolder(View itemView, BeaconsRecyclerViewAdapter adapter) {
            super(itemView);
            this.adapter = adapter;
        }

        @OnClick(R.id.beacon_container)
        public void expandCollapseInfo(View v) {

            // Check for an expanded view, collapse if you find one
            if (adapter.expandedPosition >= 0) {
                adapter.notifyItemChanged(adapter.expandedPosition);
            }

            if (adapter.expandedPosition == getAdapterPosition()) {
                adapter.notifyItemChanged(adapter.expandedPosition);
                adapter.expandedPosition = -1;
            } else {
                // Set the current position to "expanded"
                adapter.expandedPosition = getAdapterPosition();
                adapter.notifyItemChanged(getAdapterPosition());
            }
        }
    }

    public static class EddystoneUidHolder extends BaseHolder {
        @BindView(R.id.namespace_id) TextView namespaceId;
        @BindView(R.id.instance_id) TextView instanceId;

        public EddystoneUidHolder(View itemView, BeaconsRecyclerViewAdapter adapter) {
            super(itemView, adapter);
            ButterKnife.bind(this, itemView);
        }
    }

    public static class EddystoneUrlHolder extends BaseHolder {
        @BindView(R.id.url) TextView url;

        public EddystoneUrlHolder(View itemView, BeaconsRecyclerViewAdapter adapter) {
            super(itemView, adapter);
            ButterKnife.bind(this, itemView);
        }
    }

    public static class IBeaconAltBeaconHolder extends BaseHolder {
        @BindView(R.id.proximity_uuid) TextView proximityUUID;
        @BindView(R.id.major) TextView major;
        @BindView(R.id.minor) TextView minor;

        public IBeaconAltBeaconHolder(View itemView, BeaconsRecyclerViewAdapter adapter) {
            super(itemView, adapter);
            ButterKnife.bind(this, itemView);
        }
    }

    @Override
    public int getItemCount() {
        return super.getItemCount() + 1;
    }

    @Override
    public int getItemViewType(int position) {
        if (position >= getData().size()) {
            return VIEW_TYPE_FOOTER;
        }
        BeaconSaved b = getItem(position);

        switch (b.getBeaconType()) {
            case BeaconSaved.TYPE_EDDYSTONE_UID:
                return VIEW_TYPE_EDDYSTONE_UID;
            case BeaconSaved.TYPE_EDDYSTONE_URL:
                return VIEW_TYPE_EDDYSTONE_URL;
            case BeaconSaved.TYPE_ALTBEACON:
            case BeaconSaved.TYPE_IBEACON:
                return VIEW_TYPE_IBEAON_ALTBEACON;
            default:
                return VIEW_TYPE_IBEAON_ALTBEACON;
        }
    }

    @Override
    public BaseHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case VIEW_TYPE_EDDYSTONE_UID:
                return new EddystoneUidHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.eddystone_uid_item, parent, false), this);
            case VIEW_TYPE_EDDYSTONE_URL:
                return new EddystoneUrlHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.eddystone_url_item, parent, false), this);
            case VIEW_TYPE_IBEAON_ALTBEACON:
                return new IBeaconAltBeaconHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.ibeacon_altbeacon_item, parent, false), this);
            default:
                return new BaseHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.footer_item, parent, false), this);
        }
    }

    @Override
    public void onBindViewHolder(BaseHolder holder, int position) {
        if (getItemViewType(position) != VIEW_TYPE_FOOTER) {
            BeaconSaved b = getItem(position);

            holder.moreInfo.setVisibility(expandedPosition == position ? View.VISIBLE : View.GONE);
            holder.address.setText(b.getBeaconAddress());
            holder.distance.setText(String.format(Locale.getDefault(), "%.2f", b.getDistance()));
            holder.distanceQualifier.setText(getDistanceQualifier(b.getDistance()));
            holder.rssi.setText(String.format(Locale.getDefault(), "%d", b.getRSSI()));
            holder.tx.setText(String.format(Locale.getDefault(), "%d", b.getTxPower()));
            holder.manufacturer.setText(String.format(Locale.getDefault(), "0x%04X", b.getManufacturer()));
            holder.lastSeen.setText(DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.MEDIUM, Locale.getDefault()).format(b.getLastSeen()));

            if (b.isHasTelemetryData()) {
                holder.tlmData.setVisibility(View.VISIBLE);
                holder.battery.setText(String.format(Locale.getDefault(), "%d", b.getBatteryMilliVolts()));
                holder.pduCount.setText(CountHelper.coolFormat(b.getPduCount(), 0));
                holder.uptime.setText(CountHelper.coolFormat(b.getUptime(), 0));
                holder.temperature.setText(String.format(Locale.getDefault(), "%.1f", b.getTemperature()));
            } else {
                holder.tlmData.setVisibility(View.GONE);
            }

            switch (getItemViewType(position)) {
                case VIEW_TYPE_EDDYSTONE_UID:
                    EddystoneUidHolder eddyUid = (EddystoneUidHolder) holder;

                    eddyUid.beaconType.setText(String.format(Locale.getDefault(), "%s%s",
                            ctx.getString(R.string.eddystone_uid),
                            b.isHasTelemetryData() ? ctx.getString(R.string.plus_tlm) : ""));
                    eddyUid.namespaceId.setText(b.getNamespaceId());
                    eddyUid.instanceId.setText(b.getInstanceId());
                    break;
                case VIEW_TYPE_EDDYSTONE_URL:
                    EddystoneUrlHolder eddyUrl = (EddystoneUrlHolder) holder;

                    eddyUrl.beaconType.setText(String.format(Locale.getDefault(), "%s%s",
                            ctx.getString(R.string.eddystone_url),
                            b.isHasTelemetryData() ? ctx.getString(R.string.plus_tlm) : ""));
                    eddyUrl.address.setText(b.getBeaconAddress());
                    eddyUrl.url.setText(b.getURL());
                    break;
                case VIEW_TYPE_IBEAON_ALTBEACON:
                    IBeaconAltBeaconHolder h = (IBeaconAltBeaconHolder) holder;

                    h.beaconType.setText(String.format(Locale.getDefault(), "%s",
                            b.getBeaconType() == BeaconSaved.TYPE_IBEACON ? ctx.getString(R.string.ibeacon) : ctx.getString(R.string.altbeacon)));
                    h.proximityUUID.setText(b.getUUID());
                    h.major.setText(b.getMajor());
                    h.minor.setText(b.getMinor());
                    break;
            }
        }
    }

    private String getDistanceQualifier(double distance) {
        if (distance < 0.5) {
            return ctx.getString(R.string.immediate);
        } else if (distance < 5) {
            return ctx.getString(R.string.near);
        } else if (distance < 20) {
            return ctx.getString(R.string.far);
        } else {
            return ctx.getString(R.string.unknown);
        }
    }
}

File: app/src/main/java/com/bridou_n/beaconscanner/features/beaconList/BeaconsRecyclerViewAdapter.kt
package com.bridou_n.beaconscanner.features.beaconList

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bridou_n.beaconscanner.R
import com.bridou_n.beaconscanner.models.BeaconSaved
import com.bridou_n.beaconscanner.models.BeaconSaved.Companion.TYPE_ALTBEACON
import com.bridou_n.beaconscanner.models.BeaconSaved.Companion.TYPE_EDDYSTONE_UID
import com.bridou_n.beaconscanner.models.BeaconSaved.Companion.TYPE_EDDYSTONE_URL
import com.bridou_n.beaconscanner.models.BeaconSaved.Companion.TYPE_RUUVITAG
import com.bridou_n.beaconscanner.utils.extensionFunctions.toCoolFormat
import com.google.android.flexbox.FlexDirection
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.flexbox.JustifyContent
import kotlinx.android.synthetic.main.item_beacon.view.*
import java.util.*

/**
 * Created by bridou_n on 30/09/2016.
 */

typealias OnControlsOpen = (beacon: BeaconSaved) -> Unit

class BeaconsRecyclerViewAdapter(
        val ctx: Context,
        val clickListener: OnControlsOpen?
) : ListAdapter<BeaconSaved, BeaconsRecyclerViewAdapter.BeaconViewHolder>(diffCallback) {

    companion object {
        val diffCallback = object : DiffUtil.ItemCallback<BeaconSaved>() {
            override fun areItemsTheSame(oldItem: BeaconSaved, newItem: BeaconSaved) = oldItem.hashcode == newItem.hashcode
            override fun areContentsTheSame(oldItem: BeaconSaved, newItem: BeaconSaved) = oldItem == newItem
        }
    }

    class BeaconViewHolder(itemView: View, val listener: OnControlsOpen?) : RecyclerView.ViewHolder(itemView) {

        private val beaconInfosAdapter = BeaconInfosRvAdapter()

        fun bindView(beacon: BeaconSaved) {
            val ctx = itemView.context
            
            itemView.header_container.setOnClickListener {
                true.also { listener?.invoke(beacon) }
            }

            itemView.beacon_type.text = ctx.getString(when (beacon.beaconType) {
                TYPE_ALTBEACON -> R.string.altbeacon
                TYPE_EDDYSTONE_UID -> R.string.eddystone_uid
                TYPE_EDDYSTONE_URL -> R.string.eddystone_url
                TYPE_RUUVITAG -> R.string.ruuvitag
                else -> R.string.ibeacon
            })
            itemView.distance.text = String.format(Locale.getDefault(), "%.2f", beacon.distance)

            itemView.address.text = beacon.beaconAddress
            itemView.manufacturer.text = String.format(Locale.getDefault(), "0x%04X", beacon.manufacturer)
            itemView.last_seen.text = DateUtils.getRelativeTimeSpanString(
                    beacon.lastSeen, Date().time,
                    DateUtils.SECOND_IN_MILLIS,
                    DateUtils.FORMAT_ABBREV_RELATIVE
            ).toString()

            val beaconInfos = mutableListOf<BeaconInfo>().apply {
                // Adding iBeacon or AltBeacon data
                beacon.ibeaconData?.let {
                    add(BeaconInfo(ctx.getString(R.string.uuid), it.uuid))
                    add(BeaconInfo(ctx.getString(R.string.major), it.major))
                    add(BeaconInfo(ctx.getString(R.string.minor), it.minor))
                }

                add(BeaconInfo(ctx.getString(R.string.RSSI_title), String.format(ctx.getString(R.string.x_dbm), beacon.rssi)))
                add(BeaconInfo(ctx.getString(R.string.TX_title), String.format(ctx.getString(R.string.x_dbm), beacon.txPower)))

                // Adding Eddystone UID data
                beacon.eddystoneUidData?.let {
                    add(BeaconInfo(ctx.getString(R.string.namespace_id_title), it.namespaceId))
                    add(BeaconInfo(ctx.getString(R.string.instance_id_title), it.instanceId))
                }

                // Adding Eddystone URL data
                beacon.eddystoneUrlData?.let {
                    add(BeaconInfo(
                            title = ctx.getString(R.string.url_title),
                            content = it.url ?: "UNKNOWN",
                            onItemClicked = { onUrlClicked(it.url) }
                    ))
                }

                // Adding RuuviTag data
                beacon.ruuviData?.let {
                    add(BeaconInfo(ctx.getString(R.string.air_pressure_title), String.format(ctx.getString(R.string.x_hpa), it.airPressure)))
                    add(BeaconInfo(ctx.getString(R.string.temperature_title), ctx.getString(R.string.x_degrees, "${it.temperatue}")))
                    add(BeaconInfo(ctx.getString(R.string.humidity_title), String.format("%d %%", it.humidity)))
                }

                // Adding TLM data
                beacon.telemetryData?.let {
                    add(BeaconInfo(ctx.getString(R.string.battery_title), String.format(ctx.getString(R.string.x_mv), it.batteryMilliVolts)))
                    add(BeaconInfo(ctx.getString(R.string.temperature_title), ctx.getString(R.string.x_degrees, String.format("%.1f", it.temperature)))) // %.1f

                    add(BeaconInfo(ctx.getString(R.string.uptime_title), String.format(ctx.getString(R.string.x_seconds), it.uptime.toCoolFormat())))
                    add(BeaconInfo(ctx.getString(R.string.pdu_title), it.pduCount.toCoolFormat()))
                }
            }

            itemView.infos_rv.apply {
                layoutManager = FlexboxLayoutManager(context).apply {
                    flexDirection = FlexDirection.ROW
                    justifyContent = JustifyContent.SPACE_BETWEEN
                }
                adapter = beaconInfosAdapter
                isNestedScrollingEnabled = false
            }
            beaconInfosAdapter.submitList(beaconInfos)
        }

        fun onUrlClicked(url: String?) {
            url ?: return

            try {
                val uri = Uri.parse(url)
                itemView.context.startActivity(Intent(Intent.ACTION_VIEW, uri))
            } catch (e: Exception) { }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BeaconViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_beacon, parent, false)

        return BeaconViewHolder(view, clickListener)
    }

    override fun onBindViewHolder(holder: BeaconViewHolder, position: Int) {
        holder.bindView(getItem(position))
    }
}
